### DISCUSSION TITLE
Theme support for extended colours and named colour in theme reference

### DISCUSSION THREAD

**Discussion by anirbanbasu** [widgetti/solara#646] [Q&A] [answered]
Hi,

I have setup my light and dark theme specifications as follows, based on the information available at https://solara.dev/documentation/components/lab/theming and the theme generator at https://theme-generator.vuetifyjs.com/.

```
solara.lab.theme.themes.light.primary = "#2196f3"
solara.lab.theme.themes.light.secondary = "#ff5722"
solara.lab.theme.themes.light.accent = "#607d8b"
solara.lab.theme.themes.light.error = "#f44336"
solara.lab.theme.themes.light.warning = "#ffc107"
solara.lab.theme.themes.light.info = "#00bcd4"
solara.lab.theme.themes.light.success = "#8bc34a"

solara.lab.theme.themes.dark.primary = "#673ab7"
solara.lab.theme.themes.dark.secondary = "#ff5722"
solara.lab.theme.themes.dark.accent = "#cddc39"
solara.lab.theme.themes.dark.error = "#f44336"
solara.lab.theme.themes.dark.warning = "#ffc107"
solara.lab.theme.themes.dark.info = "#00bcd4"
solara.lab.theme.themes.dark.success = "#8bc34a"
```

**Q1**: However, this does not specify the "on" colours discussed at https://m2.material.io/design/material-theming/implementing-your-theme.html. Are they supported? Can I specify colours such as `solara.lab.theme.themes.light.onPrimary` since `solara.lab.theme.themes.light.on-primary` is not a valid name? There are also other colours as described at https://m2.material.io/design/material-theming/implementing-your-theme.html -- are they supported?

As a consequence of lacking the ability of specifying the "on" and the other colours, with the dark mode it is impossible to see the currently selected top navigation menu item at https://huggingface.co/spaces/xtremebytes/TLDRLC.

**Q2**: If I want to refer to a particular named colour from the theme using the _style_ to specify the CSS, I could go with something like this. However, using this, the colour does not change when the theme is toggled from dark to light or light to dark. Why?

```
style {'color': solara.lab.theme.themes.dark.accent
        if solara.lab.theme.dark_effective
        else solara.lab.theme.themes.light.accent}
```

**Q3**: Finally, is Material 3 -- https://m3.material.io/ -- supported?

**Comment by iisakkirotko** [comment]
Hey @anirbanbasu!

**Q1**: It's actually unclear to me whether Vuetify supports "on" colours. I'll take a look and get back to you. However, in general I would expect `on-primary` from Vuetify to translate to `on_vuetify` in ipyvuetify.

**Q2**: This is because `solara.lab.theme.dark_effective` doesn't behave the way you expect. You should use the `solara.lab.use_dark_effective`-hook instead.

**Q3**: Not currently. This is something we've talked about among the developers before, and is something we'd like to get in at some point in the future.

**Reply by anirbanbasu** [reply]
Thanks a lot for your response, @iisakkirotko!

Regarding **Q1**, do you mean `on-primary` from Vuetify will translate to `on_primary`? Not, `on_vuetify` right? Solara maps to Vuetify 2, right? I doubt Vuetify 2 supports "on" colours.

About **Q2**, I will try `solara.lab.use_dark_effective` instead of `solara.lab.dark_effective`.

**Reply by iisakkirotko** [reply]
@anirbanbasu good catch in Q1, I meant `on_primary` indeed.

**Reply by anirbanbasu** [reply]
@iisakkirotko thanks 👍🏼 

I am going to wait for the support of the "on" colours and Material 3.

However, for now, I would really like to fix the highlighted menu colour in the dark mode as you can see at https://huggingface.co/spaces/xtremebytes/TLDRLC -- I have not set up any colour for that. The text of the currently selected navigation menu is visible on the light mode but not at all in the dark mode.

Do you know what colour `<key>`, e.g., `solara.lab.theme.themes.dark.<key>` I can specify to make the text readable? Thanks.

**Reply by anirbanbasu** [reply]
@iisakkirotko I found something interesting.

<img width="992" alt="Screenshot 2024-05-14 at 8 19 07" src="https://github.com/widgetti/solara/assets/625357/a07c9165-3920-4110-9922-869f7f95d6e4">

I found that in the `dark` mode, it uses `primary--text` (as shown above). If I change that to `secondary--text` or `accent--text` then the text becomes readable.

In the `light` mode however, this curiously switches to `white--text`! Another curious thing is the use of  `theme--dark` even in the `light` mode. See below.

<img width="992" alt="Screenshot 2024-05-14 at 8 21 07" src="https://github.com/widgetti/solara/assets/625357/03e0c28f-d4f9-441d-84fd-bd9eb4eb5e78">

Is this a Vuetify issue? Or, something that can be fixed in Solara? If it can be fixed in Solara, how about using `accent` or `secondary` colours instead of `white` or `primary`?

**Reply by iisakkirotko** [reply]
@anirbanbasu no worries.

> However, for now, I would really like to fix the highlighted menu colour in the dark mode

This is indeed a solara issue... It's caused by pretty much what you suggested. Before Solara supported dark theme, we wanted to have a colourful AppBar instead of the grey one you get by default in Vuetify light mode. We've talked about changing the default, but didn't do it since it would be a visually breaking change. For now you can work around this by means of a custom `Layout`. There are two approaches I think might be suitable:

```python
@solara.component
def Layout():
    with solara.AppLayout(toolbar_dark=use_dark_effective(), color=..., ...):
        ...
```

Where `color` can't be `"primary"`. Or

```python
@solara.component
def Layout():
    color = "white" if solara.lab.use_dark_effective() else "black"
    with solara.AppLayout(...):
        with solara.lab.Tabs(color=color, slider_color=color, ...):
            ...
```

**Reply by anirbanbasu** [reply]
@iisakkirotko thanks a lot for the options. I will try it out.

Currently, the navigation bar in my app is shown automatically using routing. See: https://github.com/anirbanbasu/tldrlc/blob/master/app.py

Thus, to make the `with solara.AppLayout(...):` work, do I need to include it in all pages? Or, maybe a better option is that I should refactor the router options as a page and setup the tabs there by myself?

**Reply by iisakkirotko** [reply]
@anirbanbasu you could include it in all pages, that way it should be picked up automatically. You could also define it in `app.py`, and pass it to the routes using `layout=Layout`.

**Reply by anirbanbasu** [reply]
@iisakkirotko Thanks. Reading the documentation at https://solara.dev/documentation/advanced/howto/layout, the following code does not work in a different way. While the colour does work correctly in the dark mode (I opted for changing the background colour of the navigation bar instead of the text colour), the navigation bar gets displayed only for the default `/` route. For all other routes, the navigation bar vanishes. Why?

```python
reactive_colour: solara.Reactive[str] = solara.reactive("")

def set_theme_colours():
    reactive_colour.value = (
        solara.lab.theme.themes.dark.secondary
        if solara.lab.use_dark_effective()
        else solara.lab.theme.themes.light.primary
    )

@solara.component
def Layout(children: Any = []):
    set_theme_colours()
    return solara.AppLayout(
        children=children,
        color=reactive_colour.value,
        navigation=True,
        sidebar_open=False,
    )

routes = [
    solara.Route(path="/", component=page1.Page, layout=Layout),
    solara.Route(path="p4", component=page2.Page, layout=Layout),
    solara.Route(path="p3", component=page3.Page, layout=Layout),
]
```

Also, I am wondering based on your example above, how should I define the tab children? Perhaps, iterate over my `routes` variable? But, wouldn't this `layout=Layout` cause some kind of circular referencing problem? 

```python
with solara.lab.Tabs(color=color, slider_color=color, ...):
     ...
```

**Reply by iisakkirotko** [reply, marked as answer]
@anirbanbasu sorry, I think I should have phrased my previous answer better. If passing `layout=Layout` to the routes, you only need to pass it to the "root path", i.e.

```python
routes = [
    solara.Route(path="/", component=page1.Page, layout=Layout),
    solara.Route(path="p4", component=page2.Page),
    solara.Route(path="p3", component=page3.Page),
]
```

You can see some more discussion about why this happens a couple comments down in [this issue](https://github.com/widgetti/solara/issues/584).

**Comment by anirbanbasu** [comment]
@iisakkirotko thanks a lot! That worked (see commit: https://github.com/anirbanbasu/tldrlc/commit/d79dcefaae05ee9f86757c82b47603e5f040dd0e). I shall mark your last reply as the answer.
